tests/autorestart_test/test_container_autorestart.py

```diff
@@ -0,0 +1,340 @@
+                           check_container_state, duthost, container_name, True)
+    if not restarted:
+        if is_hiting_start_limit(duthost, container_name):
+            logger.info("{} hits start limit and clear reset-failed flag".format(container_name))
+            duthost.shell("sudo systemctl reset-failed {}.service".format(container_name))
+            duthost.shell("sudo systemctl start {}.service".format(container_name))
+            restarted = wait_until(CONTAINER_RESTART_THRESHOLD_SECS,
+                                   CONTAINER_CHECK_INTERVAL_SECS,
+                                   check_container_state, duthost, container_name, True)
+            pytest_assert(restarted, "Failed to restart container '{}' after reset-failed was cleared".format(container_name))
+        else:
+            pytest.fail("Failed to restart container '{}'".format(container_name))
+
+    logger.info("Container '{}' was restarted".format(container_name))
+
+
+def verify_no_autorestart_with_non_critical_process(duthost, container_name, program_name,
+                                                    program_status, program_pid):
+    """
+    @summary: Kill a non-critical process in a container to verify whether the container
+              remains in the running state
+    """
+    if program_status == "RUNNING":
+        kill_process_by_pid(duthost, container_name, program_name, program_pid)
+    elif program_status in ["EXITED", "STOPPED", "STARTING"]:
+        pytest.fail("Program '{}' in container '{}' is in the '{}' state, expected 'RUNNING'"
+                    .format(program_name, container_name, program_status))
+    else:
+        pytest.fail("Failed to find program '{}' in container '{}'"
+                    .format(program_name, container_name))
+
+    logger.info("Waiting to ensure container '{}' does not stop...".format(container_name))
+    stopped = wait_until(CONTAINER_STOP_THRESHOLD_SECS,
+                         CONTAINER_CHECK_INTERVAL_SECS,
+                         check_container_state, duthost, container_name, False)
+    pytest_assert(not stopped, "Container '{}' was stopped unexpectedly".format(container_name))
+    logger.info("Container '{}' did not stop".format(container_name))
+    logger.info("Restart the program '{}' in container '{}'".format(program_name, container_name))
+    duthost.shell("docker exec {} supervisorctl start {}".format(container_name, program_name))
+
+
+def test_containers_autorestart(duthost):
+    """
+    @summary: Test the auto-restart feature of each container against two scenarios: killing
+              a non-critical process to verify the container is still running; killing each
+              critical process to verify the container will be stopped and restarted
+    """
+    container_autorestart_states = get_container_autorestart_states(duthost)
+    disabled_containers = get_disabled_container_list(duthost)
+
+    for container_name in container_autorestart_states.keys():
```

Comment:
It would better to make container name a parameter, there should be a way to make container list a parameter fixture. So that each container would have a unique test case to mark success or failure.

This change can come as a follow up PR.